### DISCUSSION TITLE
API rate limit while downloading actions in `Set up job` step.

### DISCUSSION THREAD

**Discussion by retzero** [community/community#65858] [Enterprise]
### Select Topic Area

Question

### Body

Why do you always delete the previously downloaded `_work/_actions/` directory?
https://github.com/actions/runner/blob/main/src/Runner.Worker/ActionManager.cs#L86

Background:
We're running GHES with [GitHub Connect](https://docs.github.com/en/enterprise-server@3.7/admin/github-actions/managing-access-to-actions-from-githubcom/enabling-automatic-access-to-githubcom-actions-using-github-connect). But frequently we met `api rate limit` issue while downloading actions from `github.com` in the very early stage, `Set up job`.

```
Version 2.x was not found in the local cache

Warning: Failed to download action 'https://api.github.com/repos/actions/setup-python/tarball/e9aba2c848f5ebd159c070c61ea2c4e2b122355e'. Error: The request was canceled due to the configured HttpClient.Timeout of 100 seconds elapsing.

Error: API rate limit exceeded for 207.95.42.89. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)
```

So I want to keep the previously downloaded actions. But I realized that this directory (`_work/_actions/`) always got deleted before the build starts.

Do you have any specific reason to delete this directory? Is this intended to keep actions up to date (eg, revision changes in `v3` tag)?

How can I overcome this `api rate limit` problem?

**Comment by arafajs** [comment]
There are a few reasons the Actions runner deletes the `_work/_actions` directory before each job:

1. To ensure a clean build environment and avoid accidental reuse of old action code. Actions can change over time, so deleting the directory guarantees you get the latest code.

2. To save disk space by cleaning up old action archives after each job. Without deleting, the directories could grow very large over time.

3. The Actions runner is designed to be stateless and ephemeral. Reusing old action code goes against this philosophy of rebuilding from scratch each time.

However, I understand the rate limiting issue you're running into with GitHub Enterprise Server and GitHub Connect. Here are some options to mitigate that:

- Increase the HTTP timeout in the Actions runner configuration to allow more time for downloads before timing out.

- Configure caching for the `_work/_actions` directory in your workflow. This will save them to the cache and restore for future jobs, avoiding repeated downloads.

- Use an Actions cache server/proxy that sits between the runner and GitHub to serve cached copies of actions. This avoids hitting the GitHub API limits.

- Migrate your workflows to use Docker containers or JavaScript/TypeScript actions whenever possible. These can be cached and avoid the need to download action archives.

- Rate limit the downloads on your end to stay under the GitHub API limits. Possibly download actions at the start in a staggered way.

- Contact GitHub support to see if API limits can be increased for your GitHub Enterprise Server instance.
